.travis.yml

```diff
@@ -16,8 +13,119 @@ before_cache:
-  - ./gradlew build
+    - $HOME/.gradle/caches
+    - $HOME/.gradle/wrapper
+    - $HOME/.android/build-cache
+jobs:
+  include:
+    - name: "Build and unit tests"
+      language: java
+      install: skip
+      script:
+        - ./gradlew build
+
+    - name: "C++ functional tests"
+      language: cpp
+      addons:
+        apt:
+          packages:
+            - ninja-build
+            - valgrind
+      before_script:
+        - cd examples
+      script:
+        - ./scripts/build-cpp --valgrind --buildGluecodium
+
+    - name: "Android functional tests"
+      dist: xenial
+      language: java
+      jdk: openjdk8
+      addons:
+        apt:
+          sources:
+            - ubuntu-toolchain-r-test
+          packages:
+            - g++-7
+            - ninja-build
+      before_install:
+        - export ANDROID_HOME=${HOME}/android-sdk-linux
+        - export ANDROID_NDK_HOME=${ANDROID_HOME}/ndk-bundle
+        - export SDK_ROOT=${ANDROID_HOME}
+        - export NDK_ROOT=${ANDROID_NDK_HOME}
+        - export PATH=${PATH}:${ANDROID_HOME}/tools:${ANDROID_HOME}/tools/bin:${ANDROID_HOME}/platform-tools
+        - sudo update-alternatives --install /usr/bin/g++ g++ /usr/bin/g++-7 60
+        - sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-7 60
+      install:
+        - ANDROID_API_LEVEL=android-28
+        - ANDROID_BUILD_TOOLS_VERSION=28.0.3
+        - ANDROID_CMAKE_VERSION=3.10.2.4988404
+        - wget -q "https://dl.google.com/android/repository/sdk-tools-linux-4333796.zip" -O android-sdk-tools.zip
+        - unzip -q android-sdk-tools.zip -d ${ANDROID_HOME}
+        - mkdir -p ~/.android
+        - touch ~/.android/repositories.cfg
+        - yes | sdkmanager --licenses
+        - yes | sdkmanager "platforms;${ANDROID_API_LEVEL}" > /dev/null
+        - yes | sdkmanager "build-tools;${ANDROID_BUILD_TOOLS_VERSION}" > /dev/null
+        - yes | sdkmanager "cmake;${ANDROID_CMAKE_VERSION}" > /dev/null
+      before_script:
+        - cd examples
+      script:
+        - ./scripts/build-android --hostOnly --buildGluecodium
+
+    - name: "Swift functional tests"
+      language: cpp
+      addons:
+        apt:
+          packages:
+            - ninja-build
+            - valgrind
+      before_install:
+        - export PATH=${PWD}/cmake-install:${PWD}/cmake-install/bin:${PATH}
+      install:
+        - CMAKE_VERSION=3.15.5
+        - wget -nv https://cmake.org/files/v${CMAKE_VERSION%.*}/cmake-${CMAKE_VERSION}-Linux-x86_64.tar.gz
+        - tar xf cmake-${CMAKE_VERSION}-Linux-x86_64.tar.gz > /dev/null
+        - mv cmake-${CMAKE_VERSION}-Linux-x86_64 cmake-install
+        - SWIFT_BRANCH=swift-5.1.3-release
+        - SWIFT_VERSION=swift-5.1.3-RELEASE
+        - SWIFT_PLATFORM=ubuntu16.04
+        - SWIFT_ARCHIVE_NAME=${SWIFT_VERSION}-${SWIFT_PLATFORM}
+        - SWIFT_URL=https://swift.org/builds/${SWIFT_BRANCH}/$(echo "${SWIFT_PLATFORM}" | tr -d .)/${SWIFT_VERSION}/${SWIFT_ARCHIVE_NAME}.tar.gz
```

Comment:
Would be nice to have the variables first before other commands.

Reply:
Done